app/receipt.go

```diff
@@ -0,0 +1,177 @@
+package app
+
+import (
+	"encoding/json"
+	"fmt"
+	"math"
+	"math/big"
+
+	wasmtypes "github.com/CosmWasm/wasmd/x/wasm/types"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	authsigning "github.com/cosmos/cosmos-sdk/x/auth/signing"
+	"github.com/ethereum/go-ethereum/common"
+	ethtypes "github.com/ethereum/go-ethereum/core/types"
+	"github.com/sei-protocol/sei-chain/utils"
+	evmkeeper "github.com/sei-protocol/sei-chain/x/evm/keeper"
+	evmtypes "github.com/sei-protocol/sei-chain/x/evm/types"
+	abci "github.com/tendermint/tendermint/abci/types"
+)
+
+const ShellEVMTxType = math.MaxUint32
+
+var ERC20ApprovalTopic = common.HexToHash("0x8c5be1e5ebec7d5bd14f71427d1e84f3dd0314c0f7b2291e5b200ac8c7c3b925")
+var ERC20TransferTopic = common.HexToHash("0xddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef")
+var EmptyHash = common.HexToHash("0x0")
+
+type AllowanceResponse struct {
+	Allowance sdk.Int         `json:"allowance"`
+	Expires   json.RawMessage `json:"expires"`
+}
+
+func (app *App) AddCosmosEventsToEVMReceiptIfApplicable(ctx sdk.Context, tx sdk.Tx, checksum [32]byte, response abci.ResponseDeliverTx) {
+	if response.Code > 0 {
+		return
+	}
+	wasmEvents := GetEventsOfType(response, wasmtypes.WasmModuleEventType)
+	logs := []*ethtypes.Log{}
+	for _, wasmEvent := range wasmEvents {
+		contractAddr, found := GetAttributeValue(wasmEvent, wasmtypes.AttributeKeyContractAddr)
+		if !found {
+			continue
+		}
+		// check if there is a ERC20 pointer to contractAddr
+		pointerAddr, _, exists := app.EvmKeeper.GetERC20CW20Pointer(ctx, contractAddr)
+		if exists {
```

Comment:
i'd suggest doing a `if !exists { continue }` to avoid the indentation

Reply:
we can't continue here because we'd have another branch for 721 https://github.com/sei-protocol/sei-chain/pull/1750/files#diff-8810d2793d992ba6de31afd91265ef3facc0612669de567b51dc1d8d029cf22fR104 I refactored this part into a func to avoid the big indentation instead